concat-kdf/src/lib.rs

```diff
@@ -0,0 +1,81 @@
+#![no_std]
+
+use core::fmt;
+use digest::{generic_array::typenum::Unsigned, Digest, FixedOutputReset, Update};
+
+#[cfg(feature = "std")]
+extern crate std;
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Error {
+    NoSecret,
+    ZeroOutput,
+    CounterOverflow,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        let msg = match self {
+            Error::NoSecret => "No secret was provided.",
+            Error::ZeroOutput => "Buffer for key has zero length.",
+            Error::CounterOverflow => "Requested key length is to big.",
+        };
+        f.write_str(msg)
+    }
+}
+
+#[cfg(feature = "std")]
+impl ::std::error::Error for Error {}
+
+fn chunk_number(length: usize, chunk_size: usize) -> usize {
+    let mut number = length / chunk_size;
+    if length % chunk_size != 0 {
+        number += 1;
+    }
+
+    number
+}
+
+pub fn derive_key_into<D>(secret: &[u8], other_info: &[u8], key: &mut [u8]) -> Result<(), Error>
+where
+    D: Digest + FixedOutputReset,
+{
+    if secret.len() == 0 {
+        return Err(Error::NoSecret);
+    }
+
+    if key.len() == 0 {
+        return Err(Error::ZeroOutput);
+    }
+
+    if chunk_number(key.len(), D::OutputSize::USIZE) >= (u32::MAX as usize) {
```

Comment:
This can be replaced by precalcualting
`u32::MAX * D::OutputSize::USIZE`
and comparing `key.len()` against it.
You might need to use 64 bit data types during precalculation to avoid overflows.

Reply:
Done.